CMakeLists.txt

```diff
@@ -14,112 +12,177 @@ PROJECT(NLOPT)
-CHECK_FUNCTION_EXISTS( copysign HAVE_COPYSIGN)
-
-CHECK_TYPE_SIZE(uint32_t    UINT32_T)
-IF(NOT HAVE_UINT32_T)
-  IF(MSVC)
-    SET(uint32_t "unsigned int")
-    SET(SIZEOF_UNSIGNED_INT 4)
-  ENDIF(MSVC)
-ENDIF(NOT HAVE_UINT32_T)
+option (WITH_CXX "enable cxx routines" OFF)
+option (BUILD_SHARED_LIBS "Build NLOPT as a shared library" ON)
+option (BUILD_PYTHON "build python bindings" ON)
+option (BUILD_OCTAVE "build octave bindings" ON)
+option (BUILD_GUILE "build guile bindings" ON)
+option (USE_SWIG "use SWIG to build bindings" ON)
+
+set (NLOPT_SUFFIX)
+if (WITH_CXX)
+  set (NLOPT_SUFFIX _cxx)
+endif ()
+
+if (WITH_CXX OR BUILD_PYTHON)
+  enable_language (CXX)
+endif ()
+
+
+# Offer the user the choice of overriding the installation directories
+set (INSTALL_LIB_DIR     lib${LIB_SUFFIX} CACHE PATH "Installation directory for libraries")
+set (INSTALL_BIN_DIR     bin              CACHE PATH "Installation directory for executables")
+set (INSTALL_INCLUDE_DIR include          CACHE PATH "Installation directory for header files")
+set (INSTALL_DATA_DIR    share/nlopt       CACHE PATH "Installation directory for data files")
+set (INSTALL_CMAKE_DIR   ${INSTALL_LIB_DIR}/cmake/nlopt CACHE PATH "Installation directory for cmake config files")
+
+# Make relative paths absolute (needed later on)
+foreach(p LIB BIN INCLUDE DATA CMAKE)
+    set(var INSTALL_${p}_DIR)
+    set(RELATIVE_INSTALL_${p}_DIR ${INSTALL_${p}_DIR})
+    if(NOT IS_ABSOLUTE "${${var}}")
+        set(${var} "${CMAKE_INSTALL_PREFIX}/${${var}}")
+    endif()
+endforeach()
+
+
+set (CMAKE_INSTALL_RPATH ${INSTALL_LIB_DIR})
+set (CMAKE_INSTALL_RPATH_USE_LINK_PATH TRUE)
+if(POLICY CMP0042)
+  # Set MACOSX_RPATH to ON
+  cmake_policy(SET CMP0042 NEW)
+endif()
+
+include (CheckIncludeFiles)
+include (CheckFunctionExists)
+include (CheckTypeSize)
+include (CheckCCompilerFlag)
 
 #==============================================================================
-# LOOK FOR THE VERSION NUMBER IN configure.ac
+# COMPILATION CHECKINGS and CONFIGURATION GENERATION
 #==============================================================================
+check_include_file (dlfcn.h HAVE_DLFCN_H)
+check_include_file (getopt.h HAVE_GETOPT_H)
+check_include_file (unistd.h HAVE_UNISTD_H)
+check_include_file (string.h HAVE_STRING_H)
+check_include_file (strings.h HAVE_STRINGS_H)
+check_include_file (inttypes.h HAVE_INTTYPES_H)
+check_include_file (memory.h HAVE_MEMORY_H)
+check_include_file (stdlib.h HAVE_STDLIB_H)
+check_include_file (stdint.h HAVE_STDINT_H)
+check_include_file (time.h HAVE_TIME_H)
+check_include_file (sys/types.h HAVE_SYS_TYPES_H)
+check_include_file (sys/stat.h HAVE_SYS_STAT_H)
+check_include_file (sys/time.h HAVE_SYS_TIME_H)
+if (HAVE_TIME_H AND HAVE_SYS_TIME_H)
+  set (TIME_WITH_SYS_TIME TRUE)
+endif ()
+check_function_exists (getpid HAVE_GETPID)
+check_function_exists (syscall HAVE_GETTID_SYSCALL)
+check_function_exists (isinf HAVE_ISINF)
+check_function_exists (isnan HAVE_ISNAN)

```

Comment:
On some systems, I vaguely recall that isinf/isnan are macros rather than functions.  Does `check_function_exists` handle this case?